libsql-sqlite3/src/vector.c

```diff
@@ -0,0 +1,558 @@
+  }
+  if( pVector->flags & VECTOR_FLAGS_STATIC ){
+    return;
+  }
+  sqlite3_free(pVector);
+}
+
+float vectorDistanceCos(const Vector *pVector1, const Vector *pVector2){
+  assert( pVector1->type == pVector2->type );
+  switch (pVector1->type) {
+    case VECTOR_TYPE_FLOAT32:
+      return vectorF32DistanceCos(pVector1, pVector2);
+    case VECTOR_TYPE_FLOAT64:
+      return vectorF64DistanceCos(pVector1, pVector2);
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+/*
+ * Parses vector from text representation (e.g. '[1,2,3]'); vector type must be set
+*/
+static int vectorParseSqliteText(
+  sqlite3_value *arg,
+  Vector *pVector,
+  char **pzErrMsg
+){
+  const unsigned char *pzText;
+  double elem;
+  float *elemsFloat;
+  double *elemsDouble;
+  int iElem = 0;
+  // one more extra character in order to safely print data from elBuf with printf-like method; will be set to zero later
+  char valueBuf[MAX_FLOAT_CHAR_SZ + 1];
+  int iBuf = 0;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT32 || pVector->type == VECTOR_TYPE_FLOAT64 );
+
+  if( pVector->type == VECTOR_TYPE_FLOAT32 ){
+    elemsFloat = pVector->data;
+  } else if( pVector->type == VECTOR_TYPE_FLOAT64 ){
+    elemsDouble = pVector->data;
+  }
+
+  if( sqlite3_value_type(arg) != SQLITE_TEXT ){
+    *pzErrMsg = sqlite3_mprintf("invalid vector: not a text type");
+    goto error;
+  }
+
+  pzText = sqlite3_value_text(arg);
+  if ( pzText == NULL ) return 0;
+
+  while( pzText && sqlite3Isspace(*pzText) )
+    pzText++;
+
+  if( *pzText != '[' ){
+    *pzErrMsg = sqlite3_mprintf("invalid vector: doesn't start with '['");
+    goto error;
+  }
+  pzText++;
+
+  // clear elBuf when we are ready to parse floats
+  memset(valueBuf, 0, sizeof(valueBuf));
+
+  for(; *pzText != '\0'; pzText++){
+    char this = *pzText;
+    if( sqlite3Isspace(this) ){
+      continue;
+    }
+    if( this != ',' && this != ']' ){
+      if( iBuf > MAX_FLOAT_CHAR_SZ ){
+        *pzErrMsg = sqlite3_mprintf("float too big while parsing vector: '%s'", valueBuf);
+        goto error;
+      }
+      valueBuf[iBuf++] = this;
+      continue;
+    }
+    // empty vector case: '[]'
+    if( this == ']' && iElem == 0 && iBuf == 0 ){
+      break;
+    }
+    if( sqlite3AtoF(valueBuf, &elem, iBuf, SQLITE_UTF8) <= 0 ){
+      *pzErrMsg = sqlite3_mprintf("invalid number: '%s'", valueBuf);
+      goto error;
+    }
+    if( iElem >= MAX_VECTOR_SZ ){
+      *pzErrMsg = sqlite3_mprintf("vector is larger than the maximum: (%d)", MAX_VECTOR_SZ);
+      goto error;
+    }
+    // clear only first bufidx positions - all other are zero
+    memset(valueBuf, 0, iBuf);
+    iBuf = 0;
+    if( pVector->type == VECTOR_TYPE_FLOAT32 ){
+      elemsFloat[iElem++] = elem;
+    } else if( pVector->type == VECTOR_TYPE_FLOAT64 ){
+      elemsDouble[iElem++] = elem;
+    }
+    if( this == ']' ){
+      break;
+    }
+  }
+  while( pzText && sqlite3Isspace(*pzText) )
+    pzText++;
+
+  if( pzText != NULL && *pzText != ']' ){
```

Comment:
Shouldn't *pzText be always equals to ']' here? Why is it ok for pzText to be NULL?

Reply:
Yes, we checked that `pzText` is not null early in this function - all this checks can be omitted. Will fix that.